azure-iot-device/azure/iot/device/common/mqtt_transport.py

```diff
@@ -654,3 +661,34 @@ def complete_operation(self, mid):
             else:
                 # fully expected.  QOS=1 means we might get 2 PUBACKs
                 logger.debug("No callback set for MID: {}".format(mid))
+
+    def cancel_all_operations(self):
+        """Complete all pending operations with cancellation, removing MID tracking"""
+        logger.debug("Cancelling all pending operations")
+        with self._lock:
+            # Clear pending operations
+            pending_ops = [
+                (mid, callback) for (mid, callback) in self._pending_operation_callbacks.items()
+            ]
```

Comment:
nit: could also do `pending_op_mids = list(self._pending_operation_callbacks.keys())`

Ths list() constricturo is necessary to convert the `dict_keys` iterator into a proper list.  This is annoying.

Reply:
Ah, never mind.  You need the callbacks below.  You could still remove the comprehnsion using the list() construcor -- I would have never expected the [] operator to behave differently than list().